cmake/Utils.cmake

```diff
@@ -176,9 +175,17 @@ function(xgboost_set_cuda_flags target)
 
   set_target_properties(${target} PROPERTIES
     CUDA_STANDARD 17
-    CUDA_STANDARD_REQUIRED ON
-    CUDA_SEPARABLE_COMPILATION OFF
-    CUDA_RUNTIME_LIBRARY Static)
+    CUDA_STANDARD_REQUIRED ON)
+
+  if(CMAKE_INTERPROCEDURAL_OPTIMIZATION AND DEFINED CMAKE_CUDA_DEVICE_LINK_OPTIONS_IPO)
+    set_target_properties(${target} PROPERTIES
+      INTERPROCEDURAL_OPTIMIZATION ON
+      CUDA_SEPARABLE_COMPILATION ON)
```

Comment:
Why is `CUDA_SEPARABLE_COMPILATION` tied to host LTO?

Reply:
The implementation of LTO in both nvcc and CMake builds upon separable compilation.  In CMake LTO for device code is only enabled if SEPARABLE_COMPILATION is also enabled, otherwise it only takes effect for the hose code.  It doesn't seems to be explicitly spelled out as such in either the CMake or CUDA documentation but they do both need to be set.  From an implementation standpoint, CMake will use different gencode arguments: `code=lto_XX` during the compilation phase and `sm_XX` during a separate device linking stage to generate `cuda_device.o` which is then linked with the resulting host objects.  Again probably something worth attaching a comment to for clarity.

Reply:
So really `CUDA_SEPARABLE_COMPILATION` has nothing to do with host LTO but is required for device LTO.  There's no CMake level option to enable IPO / LTO on a per-language basis; CMAKE_INTERPROCEDURAL_OPTIMISATION` sets the default for all languages.  I could add a configure-time option to control only device LTO, initialized by `CMAKE_INTERPROCEDURAL_OPTIMIZATION` and `USE_CUDA`; that would allow for the additional scenarios of host-LTO without device-LTO and device-LTO without host-LTO.  Would that be prefered?

Reply:
So working this a little more, I don't think there's a way to only enable device lto without host lto but it can be disabled giving valid configurations for lto of host-only, host and device, or neither.

Reply:
I don't think that's necessary. Just couldn't find documents on what you are currently using and a bit concerned on using lesser known features or relying on other project's internal behaviour.

Reply:
As far as the changes with CUDA, I'm quite happy to see changes for using the `cmake_cuda_architecture` along with setting the new policy for CMake, these are well-documented features and I assume the upstream has tests for them to keep the behaviour stable.

> giving valid configurations for lto of host-only, host and device, or neither.

I'm OK with the host only option if we can leave out the un-documented behavior of CMake. We can open an issue to keep track of enabling CUDA version. I was told that separable compilation has performance implication on the final code (for the worse in some cases), so I'm a bit hesitant to enable it without proper benchmarking first.

Reply:
> As far as the changes with CUDA, I'm quite happy to see changes for using the `cmake_cuda_architecture` along with setting the new policy for CMake, these are well-documented features and I assume the upstream has tests for them to keep the behaviour stable.

I had to make a slight change to not use `cmake_dependent_option` for it since apparently it forces the value into a `BOOL` which is obviously not what we want here.


> > giving valid configurations for lto of host-only, host and device, or neither.
> 
> I'm OK with the host only option if we can leave out the un-documented behavior of CMake. We can open an issue to keep track of enabling CUDA version. I was told that separable compilation has performance implication on the final code (for the worse in some cases), so I'm a bit hesitant to enable it without proper benchmarking first.

I updated the LTO commit so the three different options work accordingly.  In general the nvcc docs show a runtime performance hit when enabling separable compilation because the optimization passes don't look across translation units.  But the way it's implemented here is to only enable separable compilation with device LTO, which is required for it to function.  So without device LTO enabled separable compilation is still turned off.